src/redisearch_rs/trie_rs/src/ffi.rs

```diff
@@ -0,0 +1,553 @@
+#![allow(non_camel_case_types, non_snake_case)]
+
+use std::ffi::{c_char, c_int, c_void};
+
+/// Holds the length of a key string in the trie.
+///
+/// C equivalent:
+/// ```c
+/// typedef uint16_t tm_len_t;
+/// ```
+pub type tm_len_t = u16;
+
+/// This special pointer is returned when TrieMap_Find cannot find anything.
+///
+/// C equivalent:
+/// ```c
+/// void *TRIEMAP_NOTFOUND = "NOT FOUND";
+/// ```
+#[unsafe(no_mangle)]
+#[used]
+pub static mut TRIEMAP_NOTFOUND: *mut ::std::os::raw::c_void = c"NOT FOUND".as_ptr() as *mut _;
+
+/// Used by TrieMapIterator to determine type of query.
+///
+/// C equivalent:
+/// ```c
+/// typedef enum {
+///     TM_PREFIX_MODE = 0,
+///     TM_CONTAINS_MODE = 1,
+///     TM_SUFFIX_MODE = 2,
+///     TM_WILDCARD_MODE = 3,
+///     TM_WILDCARD_FIXED_LEN_MODE = 4,
+///   } tm_iter_mode;
+#[repr(C)]
+#[allow(dead_code)]
+enum tm_iter_mode {
+    TM_PREFIX_MODE = 0,
+    TM_CONTAINS_MODE = 1,
+    TM_SUFFIX_MODE = 2,
+    TM_WILDCARD_MODE = 3,
+    TM_WILDCARD_FIXED_LEN_MODE = 4,
+}
+
+/// Default iteration mode for [`TrieMap_Iterate`].
+#[unsafe(no_mangle)]
+static TM_ITER_MODE_DEFAULT: tm_iter_mode = tm_iter_mode::TM_PREFIX_MODE;
+
+/// Opaque type TrieMap. Can be instantiated with [`NewTrieMap`].
+pub enum TrieMap {}
+
+/// Callback type for passing to [`TrieMap_IterateRange`].
+///
+/// C equivalent:
+/// ```c
+/// typedef void(TrieMapRangeCallback)(const char *, size_t, void *, void *);
+/// ```
+type TrieMapRangeCallback =
+    Option<unsafe extern "C" fn(*const c_char, libc::size_t, *mut c_void, *mut c_void)>;
+
+/// Type of the functions [`TrieMapIterator_Next`], [`TrieMapIterator_NextContains`],
+/// and [`TrieMapIterator_NextWildcard`].
+///
+///  C equivalent:
+/// ```c
+/// typedef int (*TrieMapIterator_NextFunc)(TrieMapIterator *it, char **ptr, tm_len_t *len, void **value);
+/// ```
+#[expect(dead_code)]
+type TrieMapIterator_NextFunc = Option<
+    unsafe extern "C" fn(
+        it: *mut TrieMapIterator,
+        ptr: *mut *mut c_char,
+        len: *mut tm_len_t,
+        value: *mut *mut c_void,
+    ) -> c_int,
+>;
+
+/// Opaque type TrieMapIterator. Obtained from calling [`TrieMap_Iterate`].
+enum TrieMapIterator {}
```

Comment:
Needs `#[repr(C)]` as well?

Reply:
That [wouldn't compile](https://play.rust-lang.org/?version=stable&mode=debug&edition=2024&gist=6a717007222b03a1ac575e4039fe5353) though. Also, we just need `* mut TrieMapIterator` to be FFI-safe, not `TrieMapIterator` itself. But the layout of pointers _is_ defined. So in this case, no need.